src/consul.rs

```diff
@@ -91,4 +102,58 @@ impl ConsulClient {
         println!("Dropped Consul lock");
         Ok(())
     }
+
+    /// Retrieves a list of all registered services and parses their tags into DnsTag
+    pub async fn fetch_service_tags(&self) -> Result<Vec<DnsTag>, anyhow::Error> {
+        let mut services_url = self.kv_api_base_url.join("catalog/")?.join("services")?;
+
+        // Set dc if it is provided in the config
+        if let Some(dc) = &self.datacenter {
+            services_url.query_pairs_mut().append_pair("dc", dc);
+        }
+
+        // Add a filter to only match "normal" Consul services
+        services_url
+            .query_pairs_mut()
+            .append_pair("filter", "ServiceKind==\"\"");
+
+        let response = self
+            .http_client
+            .get(services_url)
+            .send()
+            .await?
+            .error_for_status()?
+            .json::<HashMap<String, Vec<String>>>()
+            .await?;
+
+        let mut dns_tags = Vec::new();
+        for (_service_name, tags) in response {
+            if let Some(dns_tag) = parse_dns_tags(tags) {
+                dns_tags.push(dns_tag);
+            }
+        }
+
+        Ok(dns_tags)
+    }
+}
+
+fn parse_dns_tags(tags: Vec<String>) -> Option<DnsTag> {
+    let mut tag_map: HashMap<String, String> = HashMap::new();
+    for tag in tags.iter().filter(|t| t.starts_with("external-dns.")) {
+        let parts: Vec<&str> = tag.split('=').collect();
+        if parts.len() == 2 {
+            tag_map.insert(parts[0].to_string(), parts[1].to_string());
+        }
+    }
+
+    if tag_map.get("external-dns.enabled") == Some(&"false".to_string()) {
+        return None;
+    }
+
+    Some(DnsTag {
+        hostname: tag_map.get("external-dns.hostname").cloned()?,
+        type_: tag_map.get("external-dns.type").cloned()?,
+        ttl: tag_map.get("external-dns.ttl").and_then(|t| t.parse().ok()),
+        value: tag_map.get("external-dns.value").cloned()?,
+    })
```

Comment:
Then this needs to be something like

```rust
    let mut records = Vec::new();
    for (identifier, mut tags) in dns_tags {
        let Some(hostname) = tags.remove("hostname") else {
            // TODO: Log that `identifier` is missing this tag.
            continue;
        };
        let Some(type_) = tags.remove("type") else {
            // TODO: Log that `identifier` is missing this tag.
            continue;
        };
        let ttl = match tags.remove("ttl").map(|t| t.parse()) {
            None => None,
            Some(Ok(ttl)) => Some(ttl),
            Some(Err(e)) => {
                // TODO: Log parse error.
                continue;
            }
        };
        let Some(value) = tags.remove("value") else {
            // TODO: Log that `identifier` is missing this tag.
            continue;
        };

        records.push(DnsTag {
            hostname,
            type_,
            ttl,
            value,
        });
    }
```

Reply:
@kyrias, you are right; an identifier would be nice to log. 

Do we need an extra field for the identifier in the tag? I think we can use service_name as the identifier.

Like this

    ```rust

    pub async fn fetch_service_tags(&self) -> Result<Vec<DnsTag>, anyhow::Error> {
        let mut services_url = self.kv_api_base_url.join("catalog/")?.join("services")?;

        // Set dc if it is provided in the config
        if let Some(dc) = &self.datacenter {
            services_url.query_pairs_mut().append_pair("dc", dc);
        }

        // Add a filter to only match "normal" Consul services
        services_url.query_pairs_mut().append_pair(
            "filter",
            r#"ServiceKind == "" and ServiceTags contains "external-dns.enable=true""#,
        );

        let response = self
            .http_client
            .get(services_url)
            .send()
            .await?
            .error_for_status()?
            .json::<HashMap<String, Vec<String>>>()
            .await?;

        let mut dns_tags = Vec::new();
        for (service_name, tags) in response {
            if let Some(dns_tag) = parse_dns_tags(service_name,tags) {
                dns_tags.push(dns_tag);
            }
        }

        Ok(dns_tags)
    }


fn parse_dns_tags(service_name:String, tags: Vec<String>) -> Option<DnsTag> {
    const PREFIX: &'static str = "external-dns.";
    // Parse service tags of the format `external-dns.<field>=<value>`.
    let mut dns_tags: HashMap<String, HashMap<String, String>> = HashMap::new();
    for tag in tags.into_iter() {
        let Some(rest) = tag.strip_prefix(PREFIX) else {
            continue;
        };
        let Some((field, value)) = rest.split_once('=') else {
            continue;
        };
        dns_tags
            .entry(service_name.to_string())
            .or_default()
            .insert(field.to_string(), value.to_string());
    }

    for (identifier, mut tags) in dns_tags {
        let Some(hostname) = tags.remove("hostname") else {
            println!("Missing hostname for identifier: {}", identifier);
            continue;
        };
        let Some(type_) = tags.remove("type") else {
            println!("Missing type for identifier: {}", identifier);
            continue;
        };
        let ttl = match tags.remove("ttl").map(|t| t.parse()) {
            None => None,
            Some(Ok(ttl)) => Some(ttl),
            Some(Err(e)) => {
                eprintln!("Failed to parse TTL for identifier {}: {}", identifier, e);
                continue;
            }
        };
        let Some(value) = tags.remove("value") else {
            println!("Missing value for identifier: {}", identifier);
            continue;
        };

        return Some(DnsTag {
            identifier,
            hostname,
            type_,
            ttl,
            value,
        });
    }

    None
}
```

Reply:
> Do we need an extra field for the identifier in the tag? I think we can use service_name as the identifier.

As per https://github.com/codercengiz/nomad-external-dns/pull/1#discussion_r1583386270, no we cannot, we need to be able to parse multiple records from the same service.